packages/core/src/interpreter.ts

```diff
@@ -568,30 +578,82 @@ export class Interpreter<
       return this;
     }
 
-    [...this.state.configuration]
-      .sort((a, b) => b.order - a.order)
-      .forEach((stateNode) => {
-        for (const action of stateNode.definition.exit) {
-          this.exec(action, this.state);
-        }
-      });
-
-    // Stop all children
-    this.children.forEach((child) => {
-      if (isFunction(child.stop)) {
-        child.stop();
-      }
-    });
+    this.initialized = false;
+    this.status = InterpreterStatus.Stopped;
 
-    // Cancel all delayed events
+    // we are going to stop within the current sync frame
+    // so we can safely just cancel this here as nothing async should be fired anyway
     for (const key of Object.keys(this.delayedEventsMap)) {
       this.clock.clearTimeout(this.delayedEventsMap[key]);
     }
 
+    // clear everything that might be enqueued
     this.scheduler.clear();
-    this.initialized = false;
-    this.status = InterpreterStatus.Stopped;
-    registry.free(this.sessionId);
+
+    // at the same time let what is currently processed to be finished
+    this.scheduler.schedule(() => {
+      const _event = toSCXMLEvent({ type: 'xstate.stop' }) as any;
+
+      const nextState = serviceScope.provide(this, () => {
+        const exitActions = flatten(
+          [...this.state.configuration]
+            .sort((a, b) => b.order - a.order)
+            .map((stateNode) =>
+              toActionObjects(
+                stateNode.onExit,
+                this.machine.options.actions as any
+              )
+            )
+        );
+
+        const [resolvedActions, updatedContext] = resolveActions(
+          this.machine as any,
+          this.state,
+          this.state.context,
+          _event,
+          exitActions,
+          this.machine.config.preserveActionOrder
+        );
+
+        const newState = new State<TContext, TEvent, TStateSchema, TTypestate>({
+          value: this.state.value,
+          context: updatedContext,
+          _event,
+          _sessionid: this.sessionId,
+          historyValue: undefined,
+          history: this.state,
+          // this filtering is somewhat questionable and we might reconsider this in the future
+          // we definitely don't want to execute raised events as the stop event is the last thing that should be processed by this machine
+          // the question is if we should allow sending events from here to other actors *while* preventing the parent that has stopped this machine from receiving an event from here
```

Comment:
I think we should allow events to be sent to other machines from here.

If the recipient is the parent, it should ignore all events after stopping (I _think_):

1. Parent starts "stop procedure" (🛑  no events can be received from this point)
2. Child receives `xstate.stop` event (🛑  no events can be received from this point)
3. Final actions are executed, which may include sending to parent
  - Any event sent to parent as an action is ignored _by parent_ (child doesn't handle this logic)
4. Procedure finished; machines terminated

Reply:
> Any event sent to parent as an action is ignored by parent (child doesn't handle this logic)

Right, this is somewhat hard to handle right now though - I will take another look at this to see if I can make it happen.

Reply:
I've managed to cover this use case, however, the implementation is slightly different. I'm filtering what is being sent out from a done/stopped machine (where the original suggestion was to reject the events on the receiver's end). I couldn't find a clear way in the parent to check if the received event was sent by one of its children. I don't think this makes any practical difference though